static/js/components/Editor.js

```diff
@@ -0,0 +1,262 @@
+    }
+  ]
+]
+
+type Props = {
+  onChange: (serialized: string) => void,
+  initialValue?: string,
+  autoFocus?: boolean,
+  dispatch: Dispatch<*>
+}
+
+type State = {
+  editorState: Object,
+  showLinkMenu: boolean
+}
+
+export class Editor extends React.Component<Props, State> {
+  node: { current: null | React$ElementRef<typeof HTMLDivElement> }
+  view: Object
+  uuid: string
+
+  constructor(props: Props) {
+    super(props)
+    this.node = React.createRef()
+    this.uuid = makeUUID(10)
+
+    const { initialValue } = props
+    const editorStateOptions = {
+      doc:     defaultMarkdownParser.parse(initialValue || ""),
+      plugins: exampleSetup({
+        schema,
+        menuBar: false
+      })
+    }
+
+    this.state = {
+      editorState:  EditorState.create(editorStateOptions),
+      showLinkMenu: false
+    }
+  }
+
+  componentDidMount() {
+    const { autoFocus } = this.props
+    const { editorState } = this.state
+
+    if (this.node.current) {
+      this.view = new EditorView(this.node.current, {
+        state:               editorState,
+        dispatchTransaction: this.dispatchTransaction
+      })
+
+      if (autoFocus) {
+        this.view.focus()
+      }
+    }
+  }
+
+  dispatchTransaction = (transaction: Object) => {
+    // we need our own handler for this so we can hook into
+    // the EditorState lifecycle to synchronize it with the
+    // redux store (via the markdown serializer)
+    const { onChange } = this.props
+
+    const state = this.view.state.apply(transaction)
+    this.view.updateState(state)
+    this.setState({ editorState: state })
+
+    onChange(defaultMarkdownSerializer.serialize(this.view.state.doc))
+  }
+
+  menuClickHandlerCreator = (command: Function) => (e: Object) => {
+    e.preventDefault()
+    command(this.view.state, this.view.dispatch, this.view)
+    this.view.focus()
+  }
+
+  renderMenuBar = () => {
+    const selectEmpty = this.view ? this.view.state.selection.empty : true
+
+    return (
+      <div className="menu-bar">
+        {menuBarManifest.map((buttonGroup, idx) => (
+          <div className="button-group" key={idx}>
+            {buttonGroup.map(({ command, title, icon }) => (
+              <div
+                className="menu-button"
+                onClick={this.menuClickHandlerCreator(command)}
+                key={title}
+              >
+                {icon}
+              </div>
+            ))}
+          </div>
+        ))}
+        <div className="button-group">
+          <div
+            className={`menu-button ${selectEmpty ? "disabled" : ""}`}
+            onClick={selectEmpty ? null : this.openLinkMenu}
+          >
+            <i className="material-icons insert_link">insert_link</i>
+          </div>
+        </div>
+      </div>
+    )
+  }
+
+  renderLinkMenu = () => {
+    return <div className="link-menu">LINK</div>
+  }
+
+  formHelpers = () => {
+    const { getForm, actionCreators } = configureForm(this.uuid, () => ({
```

Comment:
This is called in a few places and the values shouldn't be changing, but there's some overhead involved here with generating these new functions. Thoughts on maybe moving this into the constructor or at least `componentDidMount` to only incur this overhead once per component lifecycle?

Reply:
Actually, on second thought, thoughts on splitting this and the `uuid` thing out into a separate HOC that wraps `withForm`? That way you can push some of this form-specific code out of your implementation here.

Reply:
hm, I hadn't thought of that but it could make this a little more compact, I'll give it a go and see how I do, haha

Reply:
I think after looking into this a little bit I'm going to just go with putting the call to `configureForm` in the constructor, just to keep things simple